techmatters/hrm
# Create Audits table + audit_trigger

Primary reviewer: @stephenhand 

## Description
This PR:
- Updates the test `dump.sql` to reflect the state of the DB prior to this PR.
- Adds `hrm-service/migrations/20220415151254-add-updatedBy.js`: a migration to add `updatedBy` column to `Contacts` and `Cases` tables.
- Cases: `createCase` and `updateCase` functions (`hrm-service/src/case/case.ts`) set `updatedBy` column (`null` and current authorized user respectively).
- Contacts: `createContact` function (`hrm-service/src/controllers/contact-controller.js`) set `updatedBy` column (`null`).
- Adds `hrm-service/migrations/20220415170355-create-Audits.js`: a migration that:
  - Creates `Audits` table.
  - Creates `audit_trigger()` function.
  - Creates `Cases_audit_trigger`, `CaseSections_audit_trigger` and `Contacts_audit_trigger` to invoke `audit_trigger()` for each row update on the corresponding table.

NOTE: This PR is not removing the underlying logic of `CaseAudits`, as first I'd like to receive feedback on this approach. @stephenhand correct me if I'm wrong but we should only remove the "add CaseAudit" statement from the Cases endpoints to cleanup that, right?

The `Audits` table has the following shape (updated as [suggested in this PR](https://github.com/techmatters/hrm/pull/177#discussion_r856203283)):
![Screenshot from 2022-04-22 16-22-38](https://user-images.githubusercontent.com/15805319/164780594-ab395a32-3ded-400d-b630-e58a549d6e1b.png)


Where:
- `id`: is the primary key of the table, an auto incremental integer.
- `user`: the database user performing the action (for now it will always be `hrm`).
- `tableName`: is the name of the table on which the update was performed.
- `operation`: the operation (`INSERT`, `UPDATE`, `DELETE`) that raised the trigger.
- `oldRecord`: the `jsonb` representation of the entire record as it was prior to the update. For `INSERT`, this value is `null`.
- `newRecord`: the `jsonb` representation of the entire record as it is after the update. For `DELETE`, this value is `null`.
- `timestamp_trx`: timestamp on when the transaction started.
- `timestamp_stm`: timestamp of the statement that raised the trigger.
- `timestamp_clock`: current timestamp at the time of saving the record.

~[*]: Since we are not setting any env vars and the "new record" when this trigger executes is NULL, there is no value we can use to fill the `updatedBy` column of the audit record.~
- One possible idea to prevent the data to be `DELETE`d without capturing any information is to start doing "soft deletes" and record the `updatedBy` of the worker that "soft deleted" the record. In this case, and if we want to prevent `DELETE` operations, we could use a `BEFORE` trigger that just stops the execution of the statement.
- One thing I've been thinking of is that so far we are assuming that the DB will receive the `createdBy` and `updatedBy` values for the `INSERT` and `UPDATE` statements, but if there is some error on the app and this values are `null` in the SQL statement, they will also be `null` in the audit. To prevent this kind of scenarios we can validate that this values have to be correctly provided (and that can also be done in a trigger). 

~Things that I'm leaving out:~
~- `accountSid`. Does it makes sense to also include this as a column here?~
~- New record state: I think we can simply look at the table for that.~

Things that can be different:
~- Instead than storing the entire previous record state, we might want to only save the "diff", but for that we must implement a function in PL/SQL that "gives the recursive difference between two jsonbs" since that is not something we have out of the box.~
- Investigate on transaction state variables to have the `updatedBy` for the `DELETE` operations. This still depends on the client app setting the value prior to the execution of the statements, but is an idea we could try.

### Checklist
- [x] [Corresponding issue has been opened](https://bugs.benetech.org/browse/CHI-994)

### Verification steps
- Try creating `Cases`, `CaseSections` and `Contacts` records (either from Flex or directly by SQL).
- Check that each operation generates exactly one record on the `Audits` table with the expected values (as described above).
